### ISSUE TITLE
AwaitWithTimeout does not cancel its timer when condition is satisfied

### ISSUE THREAD

**Issue by cretz** [temporalio/sdk-go#986]
**Is your feature request related to a problem? Please describe.**

Right now the timer in `AwaitWithTimeout` does not get cancelled when it's no longer needed. This means an erroneous workflow task is created when timer does fire.

**Describe the solution you'd like**

See if we can compatibly start cancelling these timers. @mfateev has suggested we add a SDK version marker to the beginning of all workflows to be able to make these kinds of changes henceforth. Should be discussed.

**Comment by Quinn-With-Two-Ns** [comment]
Is there any consequences to an erroneous workflow task being created? Just me being lazy, but do we have short example that triggers this already?

**Comment by cretz** [comment]
> Is there any consequences to an erroneous workflow task being created?

Yes, lots. An otherwise idle workflow is now woken up to run. It'll be put to the top of LRU cache, replay itself, etc. We should reduce workflow tasks as much as is reasonable.

> Just me being lazy, but do we have short example that triggers this already?

We don't, but it'd be easy to do, just `workflow.AwaitWithTimeout(ctx, 5 * time.Second, func() bool { return true })` then sleep for 10 seconds and confirm via history your now-not-needed timer was fired at 5 seconds.

**Comment by cretz** [comment]
After discussion, we are going to look into a way to set some kind of SDK version on the first wft completed that works like a version marker which will let us make going-forward-only fixes to things like this.

**Comment by cretz** [comment]
This seems to be a duplicate post that was also made at https://github.com/temporalio/sdk-go/issues/1125#issuecomment-2607097161, answering there.
